### PR TITLE
[feature] ability to remove custom functions (fix #27513)

### LINKED ISSUE #27513: [Expression]Allow to remove custom functions directly from QGIS
Author Name: **Harrissou Santanna** (@DelazJ)
Original Redmine Issue: [19688](https://issues.qgis.org/issues/19688)
Affected QGIS version: 3.3(master)
Redmine category:expressions
Assignee: Nathan Woodrow

---
Using the Function Editor tab, you can create a Python function providing new expression functions to use. This function is stored in the user profile folder.
It could be nice to be able to remove unneeded functions (because they were tests eg) directly from the Editor tab, either from a contextual menu or with a provided button. Currently, you need to go to the folder in which the file is saved (which is not a common/easy place to reach - I personnaly struggle to access the AppData folder on Windows)

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Add remove button to the Functions Editor tab in the Expression builder dialog to allow removing user functions from QGIS without need to navigate to user profile directory.

Fixes #27513.

### CODE REVIEW

File: src/gui/qgsexpressionbuilderwidget.cpp [hunk 1]
```diff
@@ -422,6 +423,34 @@ void QgsExpressionBuilderWidget::btnNewFile_pressed()
   }
 }
 
+void QgsExpressionBuilderWidget::btnRemoveFile_pressed()
+{
+  QString fileName = cmbFileNames->currentItem()->text();
+  if ( QFile::remove( mFunctionsPath + QDir::separator() + fileName.append( ".py" ) ) )
+  {
+    cmbFileNames->blockSignals( true );
```
Comment (elpaso) [inline src/gui/qgsexpressionbuilderwidget.cpp:431]: You may want to use QSignalBlocker

**Comment by uclaros** [comment]
Shouldn't we have a confirmation dialog before accidentally deleting a file and bursting into tears?

File: src/gui/qgsexpressionbuilderwidget.cpp [hunk 2]
```diff
@@ -422,6 +424,35 @@ void QgsExpressionBuilderWidget::btnNewFile_pressed()
...
 
+void QgsExpressionBuilderWidget::btnRemoveFile_pressed()
+{
+  int currentRow = cmbFileNames->currentRow();
+  QString fileName = cmbFileNames->currentItem()->text();
+  if ( QFile::remove( mFunctionsPath + QDir::separator() + fileName.append( ".py" ) ) )
+  {
+    {
+      const QSignalBlocker blocker( cmbFileNames );
+      QListWidgetItem *itemToRemove = cmbFileNames->takeItem( currentRow );
```
Comment (nyalldawson) [inline src/gui/qgsexpressionbuilderwidget.cpp:435]:
```suggestion
      QListWidgetItem *itemToRemove = whileBlocking( cmbFileNames )->takeItem( currentRow );
```

Then you don't need QSignalBlocker!

Comment (alexbruy) [reply src/gui/qgsexpressionbuilderwidget.cpp]: Done, thanks!

**Comment by uclaros** [comment]
Nice!! I think *delete* would be more appropriate wording than *remove* for the confirmation message box though.

**Comment by alexbruy** [comment]
@NathanW2 may I get your opinion on this?

**Pull request merged**